File: elixir/cucumber/src/graycode.ex
defmodule :graycode do
  def given([:the, :leds, :read, input], _state, _) do
    {:ok, input}
  end

  def when_([:i, :reset, :the, :counter], _state, _) do
    {:ok, '...'}
  end

  def when_([:i, :press, :the, :button], state, _) do
    {:ok, _next(state)}
  end

  def then([:the, :leds, :should, :read, expected], state, _) do
    {expected === state, state}
  end

  defp _next(leds) do
    case leds do
      '...' -> '..o'
      '..o' -> '.oo'
      '.oo' -> '.o.'
      '.o.' -> 'oo.'
      'oo.' -> 'ooo'
      'ooo' -> 'o.o'
      'o.o' -> 'o..'
      'o..' -> '...'
    end
  end

  def main() do
    :cucumberl.run("./features/graycode.feature")
  end
end

File: elixir/cucumber/src/graycode.erl
-module(graycode).

-export([given/3, 'when'/3, then/3, main/0]).

given([the, leds, read, _Input], State, _) ->
    {ok, State}.

'when'([i, reset, the, counter], State, _) ->
    {ok, State};
'when'([i, press, the, button], State, _) ->
    {ok, State}.

then([the, leds, should, read, _Expected], State, _) ->
    {ok, State}.

main() ->
    cucumberl:run("./features/graycode.feature").
